### PR TITLE
Refactor Everything

### LINKED ISSUE #34: BetterEnd & BetterNether incompatibilities
I was making a modpack when i encountered this.

So if i have BetterX installed and use BetterX world gen
A. If tried to goto the nether, i just get stuck on the portal waiting for it to find a location to teleport me
B. If i then disable FastRandom and use the portal again, everything works and chunks generate
C. Then if i Re-enable FastRandom, I can go into the nether, but the chunks that werent generated before wont load, and if i goto an ungenerated chunk, it thread locks

Same-ish effect when going to the end but in `A.` it just thread locks. Heres the thread dump courtesy of `ModernFix` https://pastebin.com/6nVs0cri

Snip
```
"Worker-Main-21" daemon prio=4 Id=198 RUNNABLE
	at org.betterx.bclib.api.v2.generator.map.hex.HexBiomeChunk.<init>(HexBiomeChunk.java:49)
	at org.betterx.bclib.api.v2.generator.map.hex.HexBiomeMap.getChunk(HexBiomeMap.java:85)
	at org.betterx.bclib.api.v2.generator.map.hex.HexBiomeMap.getChunkBiome(HexBiomeMap.java:150)
	at org.betterx.bclib.api.v2.generator.map.hex.HexBiomeMap.getRawBiome(HexBiomeMap.java:125)
	at org.betterx.bclib.api.v2.generator.map.hex.HexBiomeMap.getBiome(HexBiomeMap.java:57)
	at org.betterx.bclib.api.v2.generator.BCLibEndBiomeSource.method_38109(BCLibEndBiomeSource.java:291)
	at net.minecraft.class_3195.method_41613(class_3195.java:147)
	at net.minecraft.class_3195.method_47931(class_3195.java:199)
	at net.minecraft.class_3195$$Lambda$39594/0x0000000804cd3d78.test(Unknown Source)
	at java.base@17.0.1/java.util.Optional.filter(Optional.java:218)
	at net.minecraft.class_3195.method_47932(class_3195.java:199)
	at net.minecraft.class_3195.method_41614(class_3195.java:125)
	at net.minecraft.class_2794.method_41044(class_2794.java:513)
	at net.minecraft.class_2794.method_41041(class_2794.java:466)
	at net.minecraft.class_2794$$Lambda$39580/0x0000000804cc9af0.accept(Unknown Source)
	at java.base@17.0.1/java.lang.Iterable.forEach(Iterable.java:75)
	at net.minecraft.class_2794.method_16129(class_2794.java:448)
	at net.minecraft.class_2806.method_39464(class_2806.java:54)
	at net.minecraft.class_2806$$Lambda$1770/0x000000080151b530.doWork(Unknown Source)
	at net.minecraft.class_2806.method_12154(class_2806.java:252)
	at net.minecraft.class_3898.method_17225(class_3898.java:684)
	at net.minecraft.class_3898$$Lambda$35748/0x0000000804785a90.apply(Unknown Source)
	at com.mojang.datafixers.util.Either$Left.map(Either.java:38)
	at net.minecraft.class_3898.method_17224(class_3898.java:673)
	at net.minecraft.class_3898$$Lambda$35601/0x0000000804756220.apply(Unknown Source)
	at java.base@17.0.1/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base@17.0.1/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at net.minecraft.class_3900.method_17634(class_3900.java:62)
	at net.minecraft.class_3900$$Lambda$35747/0x0000000804785870.run(Unknown Source)
	at net.minecraft.class_3846.method_16907(class_3846.java:91)
	at net.minecraft.class_3846.method_16900(class_3846.java:146)
	at net.minecraft.class_3846.run(class_3846.java:102)
	at java.base@17.0.1/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base@17.0.1/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base@17.0.1/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base@17.0.1/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base@17.0.1/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base@17.0.1/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

### PR DESCRIPTION
**Pull request by unilock**
This PR makes a number of opinionated changes:
- The buildscript has been redone to bring it up to speed with https://github.com/FabricMC/fabric-example-mod
- The dependency on MixinExtras has been removed, since it's not actually used
- The built-in benchmarks have been completely removed, since 100% of users will never see them, they were taking up unnecessary space, and they may be unwanted in external development environments
- The main entrypoint has been removed, since the above change reduced it to simply informational logging, which most users will also never see (the Modrinth page is a better place for this)
- The preLaunch entrypoint has been changed to check the JVM's list of available `RandomGenerator`s at the source, and throw a `RuntimeException` if it can't find the one it needs
- `RandomGeneratorRandom` now contains the `RandomGeneratorFactory` in a static field
- `RandomGeneratorRandom#split` was made to be more like that of CheckedRandom
- ~~`RandomGeneratorRandom` now implements its own `RandomSplitter` (as a private static inner class) instead of using `CheckedRandom`'s, which fixes #34~~ (upstreamed)
  - ~~`ChunkRandomMixin` has been removed accordingly~~ (upstreamed)
- ~~`RandomGeneratorRandom#nextLong` now uses `nextLong` instead of `nextInt`, oops~~ (upstreamed)
- `RandomMixin#fasterrandom$createLocalInject` now uses `ThreadLocalRandom` like the original `Random` class does
- The icon has been optimized to make it a tiny bit smaller

### CODE REVIEW

**Comment by Steveplays28** [comment]
Hi, few things. I'm not sure what AnOpenSauceDev thinks, but imo removing the built-in benchmarks isn't a good idea.
I also think the `RandomGeneratorRandom` rename is unnecessary. It was named that way to show developers that it uses `RandomGenerator` under the hood, and implements the `Random` interface.
Last but not least, removing the variables (e.x. `fabric_loader_version`) from the build script makes it harder to maintain, especially if we're going multiloader.

**Comment by AnOpenSauceDev** [comment]
>     * I'm not sure what AnOpenSauceDev thinks, but imo removing the built-in benchmarks isn't a good idea.
Minecraft could easily switch to LXM if they want, or completely overhaul their random system, so I would like to still be able to easily benchmark this.


>     * I also think the `RandomGeneratorRandom` rename is unnecessary. It was named that way to show developers that it uses `RandomGenerator` under the hood, and implements the `Random` interface.
"TheFasterRandom" might sound cool at first, but it will probably get pretty old over time.

Also, the logger message about Faster Random being present has been removed due to the main entrypoint also being removed. Not sure if it should stay or not.

Aside from that, there isn't really anything I disagree with. I'll go test this out against BetterX when i have the time.

**Comment by AnOpenSauceDev** [comment]
BetterX (I tested BetterEnd) works without the workaround mixin with this PR, and feature generation (grass, trees, etc) is now 1:1, when it wasn't previously due to `nextInt` being accidentally used in `nextLong`.

**Comment by Steveplays28** [comment]
@AnOpenSauceDev That's really interesting!
Never realized that error, probably copy pasted and forgot to swap it to `nextLong()`. Awesome that fixing that makes feature generation 1:1.
I can make a PR to fix that real quick, and move some classes into their own packages (just to clean up the main package a bit).

**Comment by Steveplays28** [comment]
Test mod sounds like a good idea, could look into that!
Maybe in the future we'd want a config for users so keeping it the same name as currently is my vote.

**Comment by davidromrell** [comment]
So, your mod benchmarking method although usually fine, in games like minecraft* doing the following "all at once" will likely be inaccurate or show thread problems instead of performance.

**Comment by AnOpenSauceDev** [comment]
This wouldn't execute all at once. 99% of the time in Java, the CPU goes through a function linearly unless you directly create a new non-blocking thread. All this should do is generate a number, make the JVM discard the result, then generate the next number. The main thread is blocked by this, so the worker threads probably aren't doing much, let alone on the same core.

If you're talking about individually timing each random function, that's also not necessary, as ideally you only want to keep one generator that's the best all-round so you don't have to waste time trying to make multiple generators work nicely together.

**Comment by davidromrell** [comment]
ok

**Comment by unilock** [comment]
@Steveplays28
> Maybe in the future we'd want a config for users so keeping it the same name as currently is my vote, even if `RandomGenerator` is just a shim.

That's a good point! I'll change the name back then.

@AnOpenSauceDev
> Is this PR still being worked on?

Sorry, I recently got a new laptop, and haven't transferred everything over from the old one yet (including my dev environment). I'll rebase it on the current master branch soon.

**Comment by Steveplays28** [comment]
Awesome. Could you check up my review comment?
Still of the opinion that template expansion for `gradle.properties` is nice to have and makes it easier to maintain the mod.

**Comment by unilock** [comment]
> Could you check up my review comment?

What review comment...?

> Still of the opinion that template expansion for `gradle.properties` is nice to have and makes it easier to maintain the mod.

The only template expansion I removed was for `loader_version`, which I think should be kept static - Faster Random doesn't strictly depend on any features added in newer versions of Fabric Loader, not even Mixin Extras.

Also, Quilt Loader doesn't always provide the latest version of Fabric Loader - so if Faster Random declares a dependency on a version of Fabric Loader that Quilt hasn't caught up to yet, it'll refuse to start in Quilt environments for no good reason.

The version I have it set to, `>=0.14`, should be adequate in most cases. It could probably even be set to `*`, but I haven't tested it on Fabric Loader versions below 0.14... not that they're used (or should be used!) anymore, anyway.

File: src/main/java/com/github/anopensaucedev/fasterrandom/FasterRandomPreLaunch.java [hunk 1]
```diff
@@ -1,26 +1,14 @@
...
 	public void onPreLaunch() {
-		LOGGER.info("Preparing to launch Faster Random!");
-		LOGGER.info("If Minecraft crashes past this point, your JVM is not supported.");
-		try{
-			Class.forName("java.util.random.RandomGeneratorFactory"); // check for JVM support.
-		}catch (ClassNotFoundException exception){
-			LOGGER.error("Your JVM does not support Faster Random! Crashing game!");
-			JOptionPane.showMessageDialog(null,"._. Faster Random has crashed... \nERROR: Your Java Virtual Machine (JVM) does not support the features that Faster Random absolutely requires to run. \nPlease get a supported JVM such as GraalVM (at https://www.graalvm.org/) or Azul JDK (at https://www.azul.com/downloads/#zulu). \nYour JVM, " + System.getProperty("java.vm.name") + " " + System.getProperty("java.vm.version")  + " Is either too old, or is not suitable for Faster Random. \nMinecraft will now crash.", "Faster Random Loader",JOptionPane.WARNING_MESSAGE);
-			throw new RuntimeException();
+		if (!RandomGeneratorFactory.all().map(RandomGeneratorFactory::name).toList().contains("L64X128MixRandom")) {
```
Comment (AnOpenSauceDev) [inline src/main/java/com/github/anopensaucedev/fasterrandom/FasterRandomPreLaunch.java:10]: Is there any reason why the pop-up was removed?

**Review by AnOpenSauceDev — APPROVED** [review]
LGTM, aside from my review question about the PreLaunch changes.

**Pull request merged**